src/Revise.jl

```diff
@@ -94,6 +94,74 @@ This list gets populated by callbacks that watch directories for updates.
 """
 const revision_queue = Set{Tuple{PkgData,String}}()
 
+const revision_event = Condition()
+
+"""
+    Revise.user_callbacks_queue
+
+Global variable, `user_callbacks_queue` holds `key` values for which the
+file has changed but the user hooks have not yet been called.
+"""
+const user_callbacks_queue = Set{Any}()
+
+const user_callbacks_by_file = Dict{String, Set{Any}}()
+const user_callbacks_by_key = Dict{Any, Any}()
+
+function add_callback(f, files, modules=nothing; key=gensym())
+    remove_callback(key)
+
+    files = map(abspath, files)
+    init_watching(files)
+
+    if modules !== nothing
+        for mod in modules
+            id = PkgId(mod)
+            pkgdata = pkgdatas[id]
+            for file in srcfiles(pkgdata)
+                absname = joinpath(basedir(pkgdata), file)
+                push!(files, absname)
+                track(mod, absname)
```

Comment:
It only recently became safe to call `track` more than once on the same file, so this seems OK. But, under what circumstances would these not already be tracked?

Reply:
I don't actually see any reason why they would already be tracked? What if someone does `Revise.add_callback([], Base) do ....` ? I'm probably missing something here.

Reply:
Ah, now I see what you mean! Yes, this is faulty code then -- I'm trying to guard for the fact that the module may not be under tracking at all yet. Under that logic, we should _first_ call `track(mod)` and only then read the data from `pkgdata`, right?

If that's the case, I should also add a doctest that tests for this correct behaviour.

Reply:
Any time you've said `using MyPkg` it should be tracked. Though perhaps the `watch_package` callback (see how it gets registered in `__init__`) has to fire first. Is that the issue?

I *think* I've make `track` safe for re-entry (see #433), but I tend to be jittery (probably, needlessly) about things getting done twice and messing up the internal accounting. Maybe poke at this a bit first and try to define more clearly what circumstances require this?

Reply:
> Any time you've said `using MyPkg` it should be tracked. 
The two things that I have in mind are:
 - packages loaded before `Revise` was loaded, e.g.
```
using MyPackage
using Revise
Revise.add_callback([], [MyPackage]) do ....
```
 - packages in the system image (e.g. `Base` or anything the user added in `PackageCompiler`).

The first one could potentially be tested by spawning another Julia process. The second one seems almost impossible to test. But in both cases, calling `Revise.track(<module>)` _before_ reading `pkgdata` seems like it's the right thing to do -- what do you think?

I wasn't thinking about a race condition related to when `watch_package` is being called; do you think we should?

Reply:
> packages loaded before Revise was loaded

If package precompilation were required for all modules, I could support that, but the existence (even in principle) of non-precompiled packages makes robust support for such a feature impossible. The problematic use case is loading the package, making some changes to the source, and then tracking with Revise: in this case Revise has no way of knowing that what's in the files does not match the code in your running session. Conversely, if you load Revise first, it "defensively" parses the entire package when it's loaded and then it has a baseline to compare any file changes. This is not necessary for precompied packages because I added a source-cache to `*.ji` files in https://github.com/JuliaLang/julia/pull/23898, and did the same for Julia itself in https://github.com/JuliaLang/julia/pull/24120. This means that you can always recover the state of the source at the time at which the `*.ji` file was built. But there is no such fix available for non-precompiled packages.

So I plan to support this feature by saying "load Revise first, otherwise you're screwed" :smile:. More seriously, I could contemplate supporting this specifically for precompiled packages, but we'd need to enforce it.

> packages in the system image (e.g. Base or anything the user added in PackageCompiler).

That's a much more important use-case. Do such files record their source text? If not, we're screwed for now but should fix that ASAP.

> But in both cases, calling Revise.track(<module>) before reading pkgdata seems like it's the right thing to do -- what do you think?

Yeah, `track` populates the structure and then you can do stuff with it.

Reply:
@ffevotte this is the thing that needs resolving: in `add_callback` we need to move the call to `Revise.track(mod)` to the place _before_ reading `pkgdatas[id]`. We don't have to call `Revise.track(mod, absname)` after that.

After that change, hopefully this is done.